File: leetcode/src/test/java/com/giwankim/leetcode/ValidPalindromeTest.java
package com.giwankim.leetcode;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.stream.Stream;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

class ValidPalindromeTest {
  ValidPalindrome sut = new ValidPalindrome();

  @ParameterizedTest
  @MethodSource
  void isPalindrome(String s, boolean expected) {
    assertThat(sut.isPalindrome(s)).isEqualTo(expected);
  }

  @ParameterizedTest
  @MethodSource("isPalindrome")
  void isPalindrome2(String s, boolean expected) {
    assertThat(sut.isPalindrome2(s)).isEqualTo(expected);
  }

  public static Stream<Arguments> isPalindrome() {
    return Stream.of(
        Arguments.of("A man, a plan, a canal: Panama", true),
        Arguments.of("race a car", false),
        Arguments.of(" ", true),
        Arguments.of("Do geese see God?", true),
        Arguments.of("Hannah", true),
        Arguments.of("Hang up!", false));
  }
}

File: leetcode-kt/src/test/kotlin/com/giwankim/leetcode/ValidPalindromeTest.kt
package com.giwankim.leetcode

import io.kotest.core.spec.style.FunSpec
import io.kotest.datatest.withTests
import io.kotest.matchers.shouldBe

class ValidPalindromeTest :
    FunSpec({
        val sut = ValidPalindrome()

        context("valid palindrome") {
            withTests(
                nameFn = { (s, expected) -> "s=${s.take(20)}..., expected=$expected" },
                ValidPalindromeCase("A man, a plan, a canal: Panama", true),
                ValidPalindromeCase("race a car", false),
                ValidPalindromeCase(" ", true),
                ValidPalindromeCase("Do geese see God?", true),
                ValidPalindromeCase("Hannah", true),
                ValidPalindromeCase("Hang up!", false),
            ) { (s, expected) ->
                sut.isPalindrome(s) shouldBe expected
            }
        }
    })

private data class ValidPalindromeCase(
    val s: String,
    val expected: Boolean,
)
